### PR TITLE
Added names and descriptions for instructions in riscv_insts_base.sail

### PR DESCRIPTION
**Pull request by jriyyya**
Implements #1 and #2 

Adds names for instructions in riscv_insts_base.sail using the attributes approach
Descriptions are added to instructions in the same file using the doc comments approach

### CODE REVIEW

**Comment by jriyyya** [comment]
Hi @ThinkOpenly, I looked at the recommended approaches you left in the comments of these issues.
I tried to implement this, please let me know if the approach is correct. If this is fine I will also try to implement Alasdair Armstrong's approach for #3

**Comment by ThinkOpenly** [comment]
> I looked at the recommended approaches you left in the comments of these issues. I tried to implement this, please let me know if the approach is correct.

You have indeed followed the approach I recommended, but you have also exposed my approach as naive. (Not your fault, surely.) The problem is the instructions which are grouped under a single `union clause ast`. Especially for the instruction name, we need a way to associate, for example, "beq" with the name "Branch if equal". This also leads to the descriptions being overly general, such that "beq" includes the descriptions for "beq", "bne", "blt", "bge", "bltu", and "bgeu". This is a structural issue in the ISA document. I know there is work being done there, but I don't know if this is part of that scope. More thought is required here.

Some of the implementations are perfect, however. The ones that have a one-to-one mapping from `union clause ast` to instruction:
- Add Immediate Word
- Fence
- Total Store Order Fence
- Instruction Fence
- Environment Call
- Machine-level Return
- Supervisor-level Return
- Environment Breakpoint
- Wait For Interrupt
- Store Fence Virtual Memory Address

How does one determine which are one-to-one mappings? The generally accurate and easy way is to look at the associated `mapping clause assembly`, on the right side of the `<->`. If there is a single mnemonic string followed immediately by "spc()", that is a one-to-one.

Possibly obvious, but I'll note that there are a great many more instructions in other `riscv_insts_*.sail` that need the same treatment.

> If this is fine I will also try to implement Alasdair Armstrong's approach for #3

#3 would be a nice one, but it's very big and I think difficult. You have been warned. ;-) I would welcome the attempt!

**Comment by jriyyya** [comment]
> Possibly obvious, but I'll note that there are a great many more instructions in other `riscv_insts_*.sail` that need the same treatment.

I am looking for some feasible and better solution for doing so

> #3 would be a nice one, but it's very big and I think difficult. You have been warned. ;-) I would welcome the attempt!

Yes, I realized that now, Though it's worth a try :p

File: model/riscv_insts_base.sail [hunk 1]
```diff
@@ -170,6 +170,25 @@ mapping clause assembly = RISCV_JALR(imm, rs1, rd)
 /* see riscv_jalr_seq.sail or riscv_jalr_rmem.sail for the execute clause. */
 
 /* ****************************************************************** */
+
+$[name branch type]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:174]: change "branch type" to "Conditional Branch"

File: model/riscv_insts_base.sail [hunk 2]
```diff
@@ -271,6 +307,20 @@ mapping clause assembly = ITYPE(imm, rs1, rd, op)
   <-> itype_mnemonic(op) ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ hex_bits_12(imm)
 
 /* ****************************************************************** */
+
+$[name shift immediate Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:311]: change name to "Shift Immediate"

File: model/riscv_insts_base.sail [hunk 3]
```diff
@@ -366,6 +426,16 @@ mapping clause assembly = RTYPE(rs2, rs1, rd, op)
   <-> rtype_mnemonic(op) ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
 
 /* ****************************************************************** */
+
+$[name Load Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:430]: change name to "Load"

File: model/riscv_insts_base.sail [hunk 4]
```diff
@@ -444,6 +514,15 @@ mapping clause assembly = LOAD(imm, rs1, rd, is_unsigned, size, aq, rl)
   <-> "l" ^ size_mnemonic(size) ^ maybe_u(is_unsigned) ^ maybe_aq(aq) ^ maybe_rl(rl) ^ spc() ^ reg_name(rd) ^ sep() ^ hex_bits_12(imm) ^ opt_spc() ^ "(" ^ opt_spc() ^ reg_name(rs1) ^ opt_spc() ^ ")"
 
 /* ****************************************************************** */
+
+$[name Store Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:518]: change name to "Store"

File: model/riscv_insts_base.sail [hunk 5]
```diff
@@ -311,6 +361,16 @@ mapping clause assembly = SHIFTIOP(shamt, rs1, rd, op)
   <-> shiftiop_mnemonic(op) ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ hex_bits_6(shamt)
 
 /* ****************************************************************** */
+
+$[name register type instruction]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:365]: I'm not sure this is helpful, so delete this line for now.

File: model/riscv_insts_base.sail [hunk 6]
```diff
@@ -229,6 +248,23 @@ mapping clause assembly = BTYPE(imm, rs2, rs1, op)
   <-> btype_mnemonic(op) ^ spc() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2) ^ sep() ^ hex_bits_13(imm)
 
 /* ****************************************************************** */
+
+$[name immediate type instruction]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:252]: I'm not sure this is helpful, so delete this line for now.

File: model/riscv_insts_base.sail [hunk 7]
```diff
@@ -517,6 +606,17 @@ mapping clause assembly = ADDIW(imm, rs1, rd)
       if sizeof(xlen) == 64
 
 /* ****************************************************************** */
+
+$[name Register-Type Word Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:610]: delete this line

File: model/riscv_insts_base.sail [hunk 8]
```diff
@@ -568,6 +668,16 @@ mapping clause assembly = RTYPEW(rs2, rs1, rd, op)
       if sizeof(xlen) == 64
 
 /* ****************************************************************** */
+$[name Shift Immediate Word Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:671]: delete "Operation"

File: model/riscv_insts_base.sail [hunk 9]
```diff
@@ -606,6 +716,26 @@ mapping clause assembly = SHIFTIWOP(shamt, rs1, rd, op)
       if sizeof(xlen) == 64
 
 /* ****************************************************************** */
+
+$[name FENCE Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:720]: change name to "Fence (memory)"

File: model/riscv_insts_base.sail [hunk 10]
```diff
@@ -706,6 +836,17 @@ mapping clause assembly = FENCE(pred, succ)
   <-> "fence" ^ spc() ^ fence_bits(pred) ^ sep() ^ fence_bits(succ)
 
 /* ****************************************************************** */
+
+$[name Total Store Order Fence Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:840]: change "Total Store Order Fence Operation" to "fence (total store order)"

File: model/riscv_insts_base.sail [hunk 11]
```diff
@@ -744,6 +885,15 @@ mapping clause assembly = FENCE_TSO(pred, succ)
   <-> "fence.tso" ^ spc() ^ fence_bits(pred) ^ sep() ^ fence_bits(succ)
 
 /* ****************************************************************** */
+$[name Instruction Fence Operation]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:888]: change name to "Fence (Instruction)"

File: model/riscv_insts_base.sail [hunk 12]
```diff
@@ -849,6 +1042,18 @@ function clause execute WFI() =
 mapping clause assembly = WFI() <-> "wfi"
 
 /* ****************************************************************** */
+$[name Store Fence Virtual Memory Address (SFENCE.VMA)]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:1045]: change name to "Store Fence (Virtual Memory Address)"

**Review by ThinkOpenly — CHANGES_REQUESTED** [review]
I think the work you've done here has merit, and adds value. There remains the larger issue of associating specific names to specific instructions, but that's a harder problem for a later time. If you fix these few simple issues up, let's get it merged. Thanks!

**Comment by jriyyya** [comment]
@ThinkOpenly  Hello, I checked your suggested changes and have applied them. Also for the one where you said the line has to be deleted, I just wanted to ensure if just the line has to be deleted or the whole description of it was also to be deleted.

File: model/riscv_insts_base.sail [hunk 13]
```diff
@@ -706,6 +833,17 @@ mapping clause assembly = FENCE(pred, succ)
   <-> "fence" ^ spc() ^ fence_bits(pred) ^ sep() ^ fence_bits(succ)
 
 /* ****************************************************************** */
+
+$[name fence (total store order)]
```
Comment (ThinkOpenly) [inline model/riscv_insts_base.sail:837]: Could you rectify this to capitalize "Fence"? You did what I originally asked, I see, but I asked for "fence" when I meant "Fence". My apologies.

**Review by ThinkOpenly — CHANGES_REQUESTED** [review]
Everything looks great, exactly as I asked! Unfortunately, one of my asks was off. Last comment, I promise, before merge! Thanks for staying with me!

File: model/riscv_insts_base.sail (hunk 13 above)
Comment (jriyyya) [reply model/riscv_insts_base.sail:837]: Oh, I should have noticed the consistency, Missed it. I have corrected it now.

**Comment by jriyyya** [comment]
Thanks for considering this PR! I think we can merge this now.

**Pull request merged**